src/variational/advi.jl

```diff
@@ -34,101 +91,125 @@ function vi(model::Model, alg::ADVI; optimizer = TruncatedADAGrad())
+    μ = randn(num_params)
+    σ = softplus.(randn(num_params))
 
-    Turing.DEBUG && @debug "Optimizing ADVI..."
-    θ = optimize(elbo, alg, q, model; optimizer = optimizer)
-    μ, ω = θ[1:length(q)], θ[length(q) + 1:end]
+    # construct variational posterior
+    d = TuringDiagMvNormal(μ, σ)
+    bs = inv.(bijector.(tuple(dists...)))
+    b = Stacked(bs, ranges)
+
+    return transformed(d, b)
+end
+
+"""
+    ADVI(samples_per_step = 1, max_iters = 1000)
+
+Automatic Differentiation Variational Inference (ADVI) for a given model.
+"""
+struct ADVI{AD} <: VariationalInference{AD}
+    samples_per_step # number of samples used to estimate the ELBO in each optimization step
+    max_iters        # maximum number of gradient steps used in optimization
+end
+
+ADVI(args...) = ADVI{ADBackend()}(args...)
+ADVI() = ADVI(1, 1000)
 
-    return MeanField(μ, ω, dists, ranges) 
+alg_str(::ADVI) = "ADVI"
+
+
+function vi(model::Model, alg::ADVI; optimizer = TruncatedADAGrad())
+    q = meanfield(model)
+    return vi(model, alg, q; optimizer = optimizer)
 end
 
-# TODO: implement optimize like this?
-# (advi::ADVI)(elbo::EBLO, q::MeanField, model::Model) = begin
-# end
+function vi(model, alg::ADVI, q::TransformedDistribution{<:TuringDiagMvNormal}; optimizer = TruncatedADAGrad())
+    Turing.DEBUG && @debug "Optimizing ADVI..."
+    # Initial parameters for mean-field approx
+    μ, σs = params(q)
+    θ = vcat(μ, invsoftplus.(σs))
+
+    # Optimize
+    optimize!(elbo, alg, q, model, θ; optimizer = optimizer)
 
-function optimize(elbo::ELBO, alg::ADVI, q::MeanField, model::Model; optimizer = TruncatedADAGrad())
-    θ = randn(2 * length(q))
+    # Return updated `Distribution`
+    return update(q, θ)
+end
+
+function vi(model, alg::ADVI, q, θ_init; optimizer = TruncatedADAGrad())
+    Turing.DEBUG && @debug "Optimizing ADVI..."
+    θ = copy(θ_init)
     optimize!(elbo, alg, q, model, θ; optimizer = optimizer)
 
+    # If `q` is a mean-field approx we use the specialized `update` function
+    if q isa Distribution
+        return update(q, θ)
+    else
+        # Otherwise we assume it's a mapping θ → q
+        return q(θ)
+    end
+end
+
+
+function optimize(elbo::ELBO, alg::ADVI, q, model, θ_init; optimizer = TruncatedADAGrad())
+    θ = copy(θ_init)
+    
+    if model isa Model
+        optimize!(elbo, alg, q, make_logjoint(model), θ; optimizer = optimizer)
+    else
+        # `model` assumed to be callable z ↦ p(x, z)
+        optimize!(elbo, alg, q, model, θ; optimizer = optimizer)
+    end
+
     return θ
 end
 
+# WITHOUT updating parameters inside ELBO
 function (elbo::ELBO)(
+    rng::AbstractRNG,
     alg::ADVI,
-    q::MeanField,
-    model::Model,
-    θ::AbstractVector{<:Real},
+    q::VariationalPosterior,
+    logπ::Function,
     num_samples
 )
-    # setup
-    varinfo = Turing.VarInfo(model)
-
-    T = eltype(θ)
-    num_params = length(q)
-    μ, ω = θ[1:num_params], θ[num_params + 1: end]
+    #   𝔼_q(z)[log p(xᵢ, z)]
```

Comment:
I suggest we re-write these comments to the docstring of this function in Markdown. Not sure if it support LaTeX though.

Reply:
So I don't think this is necessarily very useful to expose to the user. It's mainly there to explain why we do that computation internally, e.g. you come across this piece of code 1 year later and wonder why we're not subtracting `logabsdetjac` (it's a bit confusing wrt. the papers, so I messed up the sign at some point and re-did the derivation and left it there for the inevitable future when I'll get confused again:) )